File: border_islands.cpp
#include <iostream>
#include <vector>


struct point{
    uint8_t d{0};
    uint8_t k{0};

    bool keep(){
        k = d;
        return k;
    }
};

struct imap{
    size_t w=0,h=0;
    std::vector<point> points;
    void set_size(size_t W,size_t H){
        w=W;h=H;
    }
    bool boarder(size_t x,size_t y){
        return (((x==0) | (x==w-1)) | ((y==0) | (y==h-1)));
    }
    void check_neighbors(size_t x, size_t y){
        uint8_t t=points[y*w+x].keep();
        //check left
        if(t&&x>0){
            if(points[y*w+x-1].k){
                check_neighbors(x-1,y);
            }
        }
        //check right
        if(t&&x<w-1){
            if(!points[y*w+x+1].k){
                check_neighbors(x+1,y);
            }
        }
        //check down
        if(t&&y<h-1){
            if(!points[(y+1)*w+x].k){
                check_neighbors(x,y+1);
            }
        }
        if(t&&y>0){
            if(!points[(y-1)*w+x].k){
                check_neighbors(x,y-1);
            }
        }
    }
    void run2(){
        size_t x=0,y=0;
        for (x=0;x<w;x++){
            y=0;
            if(points[y*w+x].keep()){
                check_neighbors(x,y);
            }
            y=h-1;
            if(points[y*w+x].keep()){
                check_neighbors(x,y);
            }
        }
        for(y=1;y<h-1;y++){
            x=0;
            if(points[y*w+x].keep()){
                check_neighbors(x,y);
            }
            x=w-1;
            if(points[y*w+x].keep()){
                check_neighbors(x,y);
            }
        }
    }
    void run(){
        for(size_t x=0;x<w;x++){
            for(size_t y=0;y<h;y++){
                if(boarder(x,y)?points[y*w+x].keep():0){
                    check_neighbors(x,y);
                }
            }
        }
    }
    void printd(){
        for(size_t x=0;x<w;x++){
            for(size_t y=0;y<h;y++){
                std::cout<<" "<<(int)points[y*w+x].d;
            }
            std::cout<<"\n";
        }
    }

    void printk(){
        for(size_t x=0;x<w;x++){
            for(size_t y=0;y<h;y++){
                std::cout<<" "<<(int)points[y*w+x].k;
            }
            std::cout<<"\n";
        }
    }
};


int main()
{
    uint8_t data[6][6] = {
	{1,0,0,0,0,0},
	{0,1,0,1,1,1},
	{0,0,1,0,1,0},
	{1,1,0,0,1,0},
	{1,0,1,1,0,0},
	{1,0,0,0,0,1},
    };
    imap m;
    m.w = sizeof(data[0]);
    m.h = sizeof(data)/m.w;
    for(size_t x=0;x<m.w;x++){
        for(size_t y=0;y<m.h;y++){
            point temp;
            temp.d = data[y][x];
            m.points.push_back(temp);
        }
    }
    m.printd();
    m.run2();
    std::cout<<"\n\n";
    m.printk();

    m.points.clear();
    return 0;
}

File: border_islands.py
#!/usr/bin/env python

data = [
	[1,0,0,0,0,0],
	[0,1,0,1,1,1],
	[0,0,1,0,1,0],
	[1,1,0,0,1,0],
	[1,0,1,1,0,0],
	[1,0,0,0,0,1],
]

class island_map:
    w=0
    h=0
    keep = []
    
    def kp(self,x,y):
        self.keep[y*self.w+x] = data[x][y]
        return self.keep[y*self.w+x]
    
    def check_neighbors(self,x,y):
        t=self.kp(x,y)
        if t==1 and x>0:
            if self.keep[y*self.w+x]==0:
                self.check_neighbors(x-1,y)
        if t==1 and x<self.w-1:
            if self.keep[y*self.w+x+1]==0:
                self.check_neighbors(x+1,y)
        if t==1 and y>0:
            if self.keep[(y-1)*self.w+x]==0:
                self.check_neighbors(x,y-1)
        if t==1 and y<self.h-1:
            if self.keep[(y+1)*self.w+x]==0:
                self.check_neighbors(x,y+1)
        
    
    def run(self,data):
        self.w = len(data)
        self.h = len(data[0])
        for i in range(0,self.w*self.h):
            self.keep.append(0)
        j=0
        k=0
        for x in range(0,self.w):
            k=0;
            if self.kp(x,k):
                self.check_neighbors(x,k)
            k=self.h-1    
            if self.kp(x,k):
                self.check_neighbors(x,k)
        for y in range(1,self.h-1):
            j=0
            if self.kp(j,y):
                self.check_neighbors(x,k)
            j=self.w-1    
            if self.kp(j,y):
                self.check_neighbors(x,k)
                
            
    def __init__(self, data):
        self.run(data)
        
    def print_data(self):
        for x in range(len(data)):
            for y in range(len(data[0])):
                print(" " + str(data[x][y]),end='')
            print("")
            
    def print_keep(self):
        for x in range(0,self.w):
            for y in range(0,self.h):
                print(" " + str(self.keep[y*self.w+x]),end='')
            print("")

p = island_map(data)
p.print_data()
print()
p.print_keep()
